### PR TITLE
Add support for kernel lock analysis

### PR DESCRIPTION
**Pull request by TianyouLi**
As system contains more and more cores, it would be usuful to be able to analyze the kernel lock overhead that impact the overall system scalability. Here added the perfspect lock comment to collect system wide hotspot, c2c and lock contention information, that will be helpful experienced performance engineers to have a chance to look into lock related problems.

Example Command:

- ./perfspect lock
   
   collect hotspot, c2c and lock contention at the whole system in 10 seconds, with sample frequency as 11

- ./perfspect lock --duration 30 --frequency 99
  
   collect hotspot, c2c and lock contention at the whole system in 30 seconds, with sample frequency as 99

Example Output:

The output will be the perf command output with text format.

image

image

image

image

### CODE REVIEW

File: internal/script/script_defs.go [hunk 1]
```diff
@@ -958,6 +959,61 @@ fi
...
+frequency=%d
+duration=%d
+
+# collect hotspot
+perf record -F $frequency -a -g --call-graph dwarf -W -d --phys-data --sample-cpu -e cycles:pp,instructions:pp,cpu/mem-loads,ldlat=30/P,cpu/mem-stores/P -o perf_hotspot.data -- sleep $duration &
+PERF_HOTSPOT_PID=$!
+
+# collect c2c
+# perf c2c record -F $frequency -a -g --call-graph dwarf -o perf_c2c.data -- sleep $duration &
+# PERF_C2C_PID=$!
```
Comment (harp-intel) [inline internal/script/script_defs.go]: Can these three commented lines be removed?

File: internal/script/script_defs.go [hunk 2]
```diff
@@ -958,6 +959,61 @@ fi
...
+
+# collect lock
+perf lock contention -a -bv --max-stack 20 2>perf_lock_contention.txt -- sleep $duration &
+PERF_LOCK_PID=$!
+
+wait ${PERF_HOTSPOT_PID}
+# wait ${PERF_C2C_PID}
```
Comment (harp-intel) [inline internal/script/script_defs.go]: Can this commented line be removed?

File: internal/report/table_defs.go [hunk 3]
```diff
@@ -1889,3 +1902,14 @@ func codePathFrequencyTableValues(outputs map[string]script.ScriptOutput) []Fiel
 	}
 	return fields
 }
+
+func kernelLockAnalysisTableValues(outputs map[string]script.ScriptOutput) []Field {
+	fields := []Field{
+		{Name: "Hotspot without Callstack", Values: []string{sectionValueFromOutput(outputs, "perf_hotspot_no_children")}},
+		{Name: "Hotspot with    Callstack", Values: []string{sectionValueFromOutput(outputs, "perf_hotspot_callgraph")}},
+		{Name: "Cache2Cache without Callstack", Values: []string{sectionValueFromOutput(outputs, "perf_c2c_no_children")}},
+		{Name: "Cache2Cache with    CallStack", Values: []string{sectionValueFromOutput(outputs, "perf_c2c_callgraph")}},
```
Comment (harp-intel) [inline internal/report/table_defs.go]: Please remove extra spaces in "with" field names.

File: internal/report/table_defs.go [hunk 4]
```diff
@@ -606,6 +608,17 @@ var tableDefinitions = map[string]TableDefinition{
...
+	//
+	// kernel lock analysis tables
+	//
+	KernelLockAnalysisTableName: {
+		Name: KernelLockAnalysisTableName,
+		ScriptNames: []string{
+			script.ProfileKernelLockScriptName,
+		},
+		FieldsFunc: kernelLockAnalysisTableValues,
+		//HTMLTableRendererFunc: kernelLockAnalysisHTMLRenderer
```
Comment (harp-intel) [inline internal/report/table_defs.go]: Remove commented line or implement renderer.

**Review by harp-intel — CHANGES_REQUESTED** [review]
Code looks very good. I left minor suggestions inline.

- Please update README.md in the project's root directory to include the 'lock' command (similar to other commands).

- Currently, only the txt report is readable. Please add an HTML renderer for equivalent readability. 
We can consider other formats and more report format improvements in the future. Ideally, we will add Insights to help interpret the large amount of data presented to the user.

- The Lock Contention field isn't working for me. Likely due to the build/version of perf being used by PerfSpect. See tools/Makefile.
```
Lock Contention:                Error: switch `b' is not available because no BUILD_BPF_SKEL=1
 Usage: perf lock contention [<options>]

    -b, --use-bpf         use BPF program to collect lock contention stats
                          (not built-in because no BUILD_BPF_SKEL=1)
```

**Comment by TianyouLi** [comment]
@harp-intel very appreciated for your review! I will resolve all your suggestions in the coming days and add test cases accordingly. 

Regarding to the perf lock contention, you are right, the -b option might not be supported. I should check the capability and by pass the perf lock contention if the feature missing on perf. 

Thanks.

File: internal/report/table_defs.go (hunk 4 above)
Comment (TianyouLi) [reply internal/report/table_defs.go]: _done_

File: internal/report/table_defs.go (hunk 3 above)
Comment (TianyouLi) [reply internal/report/table_defs.go]: done

File: internal/script/script_defs.go (hunk 2 above)
Comment (TianyouLi) [reply internal/script/script_defs.go]: done

File: internal/script/script_defs.go (hunk 1 above)
Comment (TianyouLi) [reply internal/script/script_defs.go]: done

**Comment by TianyouLi** [comment]
You are definitely right on the report part. I'd like to further improve it based on the feedbacks and also my personal experiences. For example, it's possible to see the lock cycle overall percentage and the lock cycle distribution with call stacks. The off-cpu analysis is also very useful when combined with lock contention information. As you pointed out, it could be further improvement on the reporting part, right now we just begin with a very rough feature that useable but not quite convenient.

**Comment by harp-intel** [comment]
@TianyouLi  Well done. I appreciate the updates.

PerfSpect builds and packages a specific version of perf. This perf doesn't support the bpf option. So, let's try to make PerfSpect build and use perf that has the required bpf support.

**Comment by TianyouLi** [comment]
> @TianyouLi Well done. I appreciate the updates.
> 
> PerfSpect builds and packages a specific version of perf. This perf doesn't support the bpf option. So, let's try to make PerfSpect build and use perf that has the required bpf support.

It should work with BUILD_BPF_SKEL=1 in tools/perf makefile tools/Makefile (make LDFLAGS="-static --static"  BUILD_BPF_SKEL=1). The problem is I can not test it in my environment, I've never get the full build environment work correctly. 

I checked with the source, the current version of perf has already contains the support on BPF. But it relies on the build environment a lot to enable it. 

image

image

Since this feature could survive without perf lock contention -bv, would it be possible to have this feature merged and then deal with the perf tool issue?

**Comment by TianyouLi** [comment]
@harp-intel, on my dev box without any perfspect related , i tried the perf build with make LDFLAGS="-static --static" BUILD_BPF_SKEL=1, the following error appears. I don't think my build environment is aligned with yours.

image

**Comment by TianyouLi** [comment]
> Code looks very good. I left minor suggestions inline.
> 
> * Please update README.md in the project's root directory to include the 'lock' command (similar to other commands).
> * Currently, only the txt report is readable. Please add an HTML renderer for equivalent readability.
>   We can consider other formats and more report format improvements in the future. Ideally, we will add Insights to help interpret the large amount of data presented to the user.
> * The Lock Contention field isn't working for me. Likely due to the build/version of perf being used by PerfSpect. See tools/Makefile.
> 
> ```
> Lock Contention:                Error: switch `b' is not available because no BUILD_BPF_SKEL=1
>  Usage: perf lock contention [<options>]
> 
>     -b, --use-bpf         use BPF program to collect lock contention stats
>                           (not built-in because no BUILD_BPF_SKEL=1)
> ```

https://github.com/intel/PerfSpect/actions/runs/12229325293/job/34108928079?pr=116

It's not a hard dependencies. Would it be possible to merge this feature first?

**Review by harp-intel — APPROVED** [review]

**Pull request merged**